cumulus/pallets/collator-selection/src/lib.rs

```diff
@@ -554,6 +581,138 @@ pub mod pallet {
 			Self::deposit_event(Event::InvulnerableRemoved { account_id: who });
 			Ok(())
 		}
+
+		/// Update the candidacy bond of collator candidate `origin` to a
+		/// new amount `new_deposit`.
+		///
+		/// This call will fail if `origin` is not a collator candidate, the
+		/// updated bond is lower than the minimum candidacy bond, and/or
+		/// the amount cannot be reserved.
+		#[pallet::call_index(7)]
+		#[pallet::weight(T::WeightInfo::update_bond(T::MaxCandidates::get()))]
+		pub fn update_bond(
+			origin: OriginFor<T>,
+			new_deposit: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			ensure!(new_deposit >= <CandidacyBond<T>>::get(), Error::<T>::DepositTooLow);
+			// The function below will try to mutate the `Candidates` entry for
+			// the caller to update their deposit to the new value of
+			// `new_deposit`. The return value is a tuple of the position of
+			// the entry in the list, used for weight calculation, and the new
+			// bonded amount.
+			let (new_amount, length) = <CandidateList<T>>::try_mutate(
+				|candidates| -> Result<(BalanceOf<T>, usize), DispatchError> {
+					let mut idx = candidates
+						.iter()
+						.position(|candidate_info| candidate_info.who == who)
+						.ok_or_else(|| Error::<T>::NotCandidate)?;
+					let candidate_count = candidates.len();
+					let old_deposit = candidates[idx].deposit;
+					if new_deposit > old_deposit {
+						T::Currency::reserve(&who, new_deposit - old_deposit)?;
+					} else if new_deposit < old_deposit {
+						T::Currency::unreserve(&who, old_deposit - new_deposit);
+					}
+					candidates[idx].deposit = new_deposit;
+					if new_deposit > old_deposit && idx < candidate_count {
+						idx += 1;
+						while idx < candidate_count && candidates[idx].deposit < new_deposit {
+							candidates.as_mut().swap(idx - 1, idx);
+							idx += 1;
+						}
+					} else {
+						while idx > 0 && candidates[idx].deposit >= new_deposit {
+							candidates.as_mut().swap(idx - 1, idx);
+							idx -= 1;
```

Comment:
I think this is safe as long as `candidates` is less than `u32::MAX`, but not sure if it's still FRAME best practice to use something like `saturating_accrue`. @gpestana ?